### DISCUSSION TITLE
[Routing 6.4] Why is Route attribute class marked `@final`?

### DISCUSSION THREAD

**Discussion by zorac** [symfony/symfony#52926] [Q&A]
Hi,

I'm curious as to the reasoning behind `Symfony\Component\Routing\Attribute\Route` being annotated as `@final`. This has no impact beyond a PHPStan warning, but I'm interested to know whether that can safely be ignored, or if it's an indication that I should change my approach.

Use case: using the standalone routing component, subclassing the Route attribute primarily to provide an improved developer experience by providing a constructor which explicitly makes some parameters required and provides type-checking for custom options which are used elsewhere in the request-handling stack. Sanitised example:
```php
#[Attribute(Attribute::IS_REPEATABLE | Attribute::TARGET_METHOD)]
class CustomRoute extends Route
{
    public function __construct(
        string $path,
        string|array $methods,
        CustomOption $custom,
        array $requirements = [],
    ) {
        parent::__construct(
            path: $path,
            requirements: $requirements,
            options: ['custom' => $custom],
            methods: $methods,
        );
    }
}
```

**Comment by ghost** [comment]
For you as a developer the `@final` annotation should mean the same  what `final` keyword. 

Why annotation instead of the keyword? I don't know. Maybe Symfony needs this internally. 
  
Why final? For open source projects it's important to consider a cost of a change for non-final class. It's a BC break that cannot be introduced until next major release or a BC layer must be introduced which also needs to be maintained.

**Comment by tjveldhuizen** [comment]
I had the same question, but understand the BC-impact when the classes are not final. However, in case of attributes there is no option to decorate the class. I hope somebody has a good solution to connect my own custom route attributes to the routing component.
